### ISSUE TITLE
Review guidance on implicit typing for local variables

### ISSUE THREAD

**Issue by johannbrink** [dotnet/docs#26787]
The documentation about implicit typing for local variables only really focuses on the first part of the statement: "Use implicit typing for local variables when the type of the variable is obvious from the right side of the assignment, _or when the precise type is not important._"

Overall typed classes are part of what makes C# great, but the tooling is clever enough to figure that out even if we are using var. You can hover over a variable at any time to see what the type is.

Using explicit typing also comes at the cost of importing namespaces that you wouldn't otherwise import. We ended up with "using System.Collections.Generic" scattered throughout the codebase when using explicit typing. Having explicit types everywhere also causes the size pull requests to grow when refactoring the codebase because you have to make sweeping changes across the codebase whenever a type is renamed.

Resharper is configurable and I tested its "use var when evident" setting, but it doesn't interpret it aligned with the Microsoft documentation. If the tooling can't interpret the convention then it's probably a sign that it is confusing.

The example below assumes we use var when evident:
- var bla correctly doesn't show a warning because of the new operator
- var bookingId correctly shows a warning because there is no cast and no new operator
- var bookingIdAgain correctly doesn't show a warning because of an explicit cast
- var bookingIdAgain2 incorrectly doesn't show a warning but it should because there is no explicit cast and no new operator used.
```
var bla = new bool();
var bookingId = await HttpResponseDeserializer<Guid>.DeserializeModelAsync(placeBookingResponse);
var bookingIdAgain = (Guid)bookingId;
var bookingIdAgain2 = Guid.NewGuid();
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 40f42506-f4a5-859a-0266-ded002056d83
* Version Independent ID: f940f2ad-e13d-4d37-6439-f8afcbc4e4c7
* Content: [C# Coding Conventions](https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions)
* Content Source: [docs/csharp/fundamentals/coding-style/coding-conventions.md](https://github.com/dotnet/docs/blob/main/docs/csharp/fundamentals/coding-style/coding-conventions.md)
* Product: **dotnet-csharp**
* Technology: **csharp-fundamentals**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

---
[Associated WorkItem - 123085](https://dev.azure.com/msft-skilling/Content/_workitems/edit/123085)

**Comment by denjmpr** [comment]
Also,
It would be nice if there will be more concrete examples "when the precise type is not important.".
I spent a few hours when I enforced my collegue to change var to the actual type rely on the rule "Don't assume the type is clear from a method name.".
Then we reread the whole block and saw "when the precise type is not important." phrase and stuck.
Then he provided an example when explicit type might be really need to understand what happening:
```csharp
var e = GetValue();
return string.Format("this is {0}",e);
```
In that case we really depends of what type is e (e.g. number, which depends on culture, DBNull, DateTime, etc.).
But it means that in most other cases we can ignore other rules, e.g. mentioned "Don't assume the type is clear from a method name." and most of the time use var instead of explicit type.
